kroxylicious-systemtests/src/test/java/io/kroxylicious/systemtests/RecordEncryptionST.java

```diff
@@ -115,24 +128,126 @@ void produceAndConsumeMessage(String namespace, TestKmsFacade<?, ?, ?> testKmsFa
+        LOGGER.info("Received: {}", result);
+
+        assertThat(result).withFailMessage("expected messages have not been received!")
+                .extracting(ConsumerRecord::getValue)
+                .hasSize(numberOfMessages)
+                .allSatisfy(v -> assertThat(v).contains(MESSAGE));
+    }
+
+    @TestTemplate
+    void ensureClusterHasEncryptedMessageWithRotatedKEK(String namespace, TestKmsFacade<?, ?, ?> testKmsFacade) {
+        // Skip AWS test execution because the ciphertext blob metadata to read the version of the KEK is not available anywhere
+        assumeThat(testKmsFacade.getKmsServiceClass().getSimpleName().toLowerCase().contains("vault")).isTrue();
+        testKekManager = testKmsFacade.getTestKekManager();
+        testKekManager.generateKek("KEK_" + topicName);
+        int numberOfMessages = 1;
+        ExperimentalKmsConfig experimentalKmsConfig = new ExperimentalKmsConfig(1, 1, 1, 1);
+
+        // start Kroxylicious
+        LOGGER.info("Given Kroxylicious in {} namespace with {} replicas", namespace, 1);
+        Kroxylicious kroxylicious = new Kroxylicious(namespace);
+        kroxylicious.deployPortPerBrokerPlainWithRecordEncryptionFilter(clusterName, 1, testKmsFacade, experimentalKmsConfig);
+        bootstrap = kroxylicious.getBootstrap();
+
+        LOGGER.info("And a kafka Topic named {}", topicName);
+        KafkaSteps.createTopic(namespace, topicName, bootstrap, 1, 2);
+
+        LOGGER.info("When {} messages '{}' are sent to the topic '{}'", numberOfMessages, MESSAGE, topicName);
+        KroxyliciousSteps.produceMessages(namespace, topicName, bootstrap, MESSAGE, numberOfMessages);
+
+        LOGGER.info("Then the messages are consumed");
+        List<ConsumerRecord> resultEncrypted = KroxyliciousSteps.consumeMessageFromKafkaCluster(namespace, topicName, clusterName,
+                Constants.KAFKA_DEFAULT_NAMESPACE, numberOfMessages, Duration.ofMinutes(2));
+        LOGGER.info("Received: {}", resultEncrypted);
+
+        assertKekVersionWithinParcel(resultEncrypted, "v1", testKekManager);
+
+        LOGGER.info("When KEK is rotated");
+        testKekManager.rotateKek("KEK_" + topicName);
+
+        LOGGER.info("And {} messages '{}' are sent to the topic '{}'", numberOfMessages, MESSAGE, topicName);
+        KroxyliciousSteps.produceMessages(namespace, topicName, bootstrap, MESSAGE, numberOfMessages);
+
+        LOGGER.info("Then the messages are consumed");
+        List<ConsumerRecord> resultEncryptedRotatedKek = KroxyliciousSteps.consumeMessageFromKafkaCluster(namespace, topicName, clusterName,
+                Constants.KAFKA_DEFAULT_NAMESPACE, numberOfMessages, Duration.ofMinutes(2));
+        LOGGER.info("Received: {}", resultEncryptedRotatedKek);
+
+        assertKekVersionWithinParcel(resultEncryptedRotatedKek.stream().filter(r -> !r.getValue().contains("v1")).toList(),
```

Comment:
It'd be better to remove the records that were received earlier, then assert that the remainder have the expected value.

If you implemented ConsumerRecord#equals/hashCode, you could use:

```
resultEncryptedRotatedKek.removeAll(resultEncrypted);
```

Reply:
How about 
```suggestion
        final Set<String> versions = Set.of("v1", "v2");
        assertThat(resultEncryptedRotatedKek.stream().collect(Collectors.groupingBy(
                consumerRecord -> {
                    final String value = consumerRecord.getValue();
                    for (String key : versions) {
                        if (value.contains(key)) {
                            return key;
                        }
                    }
                    return "unknown";
                })))
                .doesNotContainKey("unknown")
                .containsKey("v2")
                .isNotEmpty();
```

I'm not sure if there is value in asserting something about `v1` or not.

Reply:
I went with the proposal from Keith. We should not check that `resultEncryptedRotatedKek` contains `v1` because it will only contained it for kaf client. The rest of the clients can get only those messages that has not been already committed.

I think removing what we have already received fits with all cases, because if it is not there, it won't remove anything.